pulsar/consumer_partition.go

```diff
@@ -683,6 +688,83 @@ func (pc *partitionConsumer) AckID(msgID MessageID) error {
 	return pc.ackID(msgID, false)
 }
 
+func (pc *partitionConsumer) AckIDList(msgIDs []MessageID) map[MessageID]error {
+	errorMap := make(map[MessageID]error)
+	if !pc.options.ackWithResponse {
+		for _, msgID := range msgIDs {
+			if err := pc.ackID(msgID, false); err != nil {
+				errorMap[msgID] = err
+			}
+		}
+		return errorMap
+	}
+
+	chunkedMsgIDs := make([]*chunkMessageID, 0) // we need to remove them after acknowledging
+	pendingAcks := make(map[position]*bitset.BitSet)
+
+	// They might be complete after the whole for loop
+	incompleteTrackingIDs := make([]*trackingMessageID, 0)
+	for _, msgID := range msgIDs {
+		if msgID.PartitionIdx() != pc.partitionIdx {
+			errorMap[msgID] = fmt.Errorf("inconsistent partition index %v (current: %v)",
+				msgID.PartitionIdx(), pc.partitionIdx)
+		} else if msgID.BatchIdx() >= 0 && msgID.BatchSize() > 0 &&
+			msgID.BatchIdx() >= msgID.BatchSize() {
+			errorMap[msgID] = fmt.Errorf("invalid batch index %v (size: %v)", msgID.BatchIdx(), msgID.BatchSize())
+		} else {
+			switch convertedMsgID := msgID.(type) {
+			case *trackingMessageID:
+				if convertedMsgID.ack() {
+					pendingAcks[newPosition(msgID)] = nil
+				} else {
+					incompleteTrackingIDs = append(incompleteTrackingIDs, convertedMsgID)
+				}
+			case *chunkMessageID:
+				for _, id := range pc.unAckChunksTracker.get(convertedMsgID) {
+					pendingAcks[newPosition(id)] = nil
+				}
+				chunkedMsgIDs = append(chunkedMsgIDs, convertedMsgID)
+			case *messageID:
+				pendingAcks[newPosition(msgID)] = nil
+			default:
+				errorMap[msgID] = fmt.Errorf("invalid message id type %T", msgID)
+			}
+		}
+	}
+
+	if pc.options.enableBatchIndexAck {
```

Comment:
A non-blocking comment for this PR:

If the `msgId` is a batch, can we record this ID in `ackGroupTracking` even when `batchIndexAck` is not enabled?

This way, even if a batch ID is not acknowledged, we can filter it out the next time the same message is received, avoiding duplicate delivery to the user.

This scenario references the issue: https://github.com/apache/pulsar/issues/23436

Reply:
It is a separated issue while this PR might be huge. Could you open an issue in this repo to track?